calypso/controller/mod.go

```diff
@@ -0,0 +1,245 @@
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"go.dedis.ch/dela-apps/calypso"
+	guictrl "go.dedis.ch/dela-apps/calypso/controller/gui/controllers"
+	"go.dedis.ch/dela/cli"
+	"go.dedis.ch/dela/cli/node"
+	"go.dedis.ch/dela/crypto/ed25519"
+	"go.dedis.ch/dela/dkg"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/mino/httpclient"
+	"golang.org/x/xerrors"
+)
+
+// formatter defines how messages are marshalled/unmarshalled for the deamon.
+// Using this variable allows us to gain flexibility for the tests.
+var formatter formatterI = jsonFormatter{}
+
+// NewMinimal returns a new minimal initializer. The static files for the client
+// GUI are not packed in the binairy, so one need to build its own binairy in
+// order to use it.
+func NewMinimal() node.Initializer {
+	return minimal{}
+}
+
+// minimal is an initializer with the minimum set of commands
+//
+// - implements node.Initializer
+type minimal struct{}
+
+// SetCommands implements node.Initializer
+func (m minimal) SetCommands(builder node.Builder) {
+	cb := builder.SetCommand("calypso")
+	cb.SetDescription("Set of commands to administrate Calypso")
+
+	sub := cb.SetSubCommand("setup")
+	sub.SetDescription("setup Calypso and create the distributed key. " +
+		"Must be run only once.")
+	sub.SetAction(builder.MakeAction(setupAction{}))
+	sub.SetFlags(
+		cli.StringFlag{
+			Name:     "pubkeys",
+			Usage:    "a list of public keys in hex strings, separated by commas",
+			Required: true,
+		},
+		cli.StringFlag{
+			Name: "addrs",
+			Usage: "a list of addresses correponding to the public keys, " +
+				"separated by commas",
+			Required: true,
+		},
+		cli.IntFlag{
+			Name:     "threshold",
+			Usage:    "the minimum number of nodes that is needed to decrypt",
+			Required: true,
+		},
+	)
+}
+
+// Inject implements node.Initializer. This function contains the initialization
+// code run on each node that wants to support Calypso. We create the dkg actor
+// and then use it to create the Calypso, which is then injected as a
+// dependency. We will need this dependency in the setup phase.
+func (m minimal) Inject(ctx cli.Flags, inj node.Injector) error {
+	var dkg dkg.DKG
+	err := inj.Resolve(&dkg)
+	if err != nil {
+		return xerrors.Errorf("failed to resolve dkg: %v", err)
+	}
+
+	actor, err := dkg.Listen()
+	if err != nil {
+		return xerrors.Errorf("failed to listen dkg: %v", err)
+	}
+
+	caly := calypso.NewCalypso(actor)
+
+	inj.Inject(caly)
+
+	var httpclient httpclient.Httpclient
+	err = inj.Resolve(&httpclient)
+	if err != nil {
+		return xerrors.Errorf("failed to resolve httpclient: %v", err)
+	}
+
+	ctrl := guictrl.NewCtrl(caly)
+
+	fs := http.FileServer(http.Dir(ctrl.Abs("gui/assets")))
+	httpclient.RegisterHandler("/assets/", tofunc(http.StripPrefix("/assets/", fs)))
+	httpclient.RegisterHandler("/", ctrl.HomeHandler())
+	httpclient.RegisterHandler("/pubkey", ctrl.PubkeyHandler())
+	httpclient.RegisterHandler("/encrypt", ctrl.EncryptHandler())
```

Comment:
Encryption cannot happen in a node as it will learn about the message. The client should get the public key and then encrypt the message with it.

Reply:
Yes, right. This functionality is more a toy for convenience and to test the app. The user provides the public key and the message, and the app encrypts it. This step can always be done separately, but the user can choose to trust the server. We do the same for the decryption of the re-encrypted message in cothority/calypso